blocks/library/video/index.js

```diff
@@ -57,116 +66,172 @@ export const settings = {
 		}
 	},
 
-	edit: class extends Component {
-		constructor( { className } ) {
-			super( ...arguments );
-			// edit component has its own src in the state so it can be edited
-			// without setting the actual value outside of the edit UI
-			this.state = {
-				editing: ! this.props.attributes.src,
-				src: this.props.attributes.src,
-				className,
-			};
-		}
+	edit: enhance( ( {
+		// Provided by `withState`:
+		editing,
+		src,
+		className,
+		hasError,
+		hasDismissedError,
+		setState,
+		// Provided by editor:
+		...props
+	} ) => {
+		console.log( 'Rendering' );
 
-		render() {
-			const { align, caption, id } = this.props.attributes;
-			const { setAttributes, isSelected } = this.props;
-			const { editing, className, src } = this.state;
-			const updateAlignment = ( nextAlign ) => setAttributes( { align: nextAlign } );
-			const switchToEditing = () => {
-				this.setState( { editing: true } );
-			};
-			const onSelectVideo = ( media ) => {
-				if ( media && media.url ) {
-					// sets the block's attribute and updates the edit component from the
-					// selected media, then switches off the editing UI
-					setAttributes( { src: media.url, id: media.id } );
-					this.setState( { src: media.url, editing: false } );
-				}
-			};
-			const onSelectUrl = ( event ) => {
-				event.preventDefault();
-				if ( src ) {
-					// set the block's src from the edit component's state, and switch off the editing UI
-					setAttributes( { src } );
-					this.setState( { editing: false } );
+		const { align, caption, id } = props.attributes;
+		const { setAttributes, isSelected } = props;
+		const updateAlignment = ( nextAlign ) => setAttributes( { align: nextAlign } );
+		const switchToEditing = () => {
+			setState( { editing: true } );
+		};
+		const onSelectVideo = ( media ) => {
+			if ( media && media.url ) {
+				// sets the block's attribute and updates the edit component from the
+				// selected media, then switches off the editing UI
+				setAttributes( { src: media.url, id: media.id } );
+				setState( { src: media.url, editing: false } );
+			}
+		};
+		const onSelectUrl = ( event ) => {
+			event.preventDefault();
+			if ( src ) {
+				// set the block's src from the edit component's state, and switch off the editing UI
+				setAttributes( { src } );
+				setState( { editing: false } );
+			}
+			return false;
+		};
+
+		const convertToEmbed = () => {
+			const { onReplace } = props;
+			if ( ! src ) {
+				return;
+			}
+			onReplace( createBlock( 'core/embed', { url: src } ) );
+		};
+
+		const tryEmbed = () => {
+			if ( ! src ) {
+				return;
+			}
+
+			const apiURL = addQueryArgs( wpApiSettings.root + 'oembed/1.0/proxy', {
+				url: src,
+				_wpnonce: wpApiSettings.nonce,
+			} );
+
+			console.log( 'tryEmbed; querying' );
+			window.fetch( apiURL, { credentials: 'include' } ).then(
```

Comment:
Should we transform this to class component to avoid calling `setState` on unmounted components?

Reply:
Either that, or we source a bound `fetch` like:

https://github.com/WordPress/gutenberg/pull/4839#issuecomment-364673030